logic/src/commonMain/kotlin/com/wire/kalium/logic/data/client/MLSClientProvider.kt

```diff
@@ -56,45 +59,61 @@ class MLSClientProviderImpl(
+        withContext(dispatchers.io) {
+            val currentClientId = clientId ?: currentClientIdProvider().fold({ return@withContext Either.Left(it) }, { it })
+            val cryptoUserId = CryptoUserID(value = userId.value, domain = userId.domain)
+            return@withContext mlsClient?.let {
+                Either.Right(it)
+            } ?: run {
+                mlsClient(
+                    cryptoUserId,
+                    currentClientId
+                ).map {
+                    mlsClient = it
+                    return@run Either.Right(it)
+                }
             }
         }
     }
 
     override suspend fun clearLocalFiles() {
-        mlsClient?.close()
-        mlsClient = null
-        FileUtil.deleteDirectory(rootKeyStorePath)
+        mlsClientMutex.withLock {
+            mlsClient?.close()
+            mlsClient = null
+            FileUtil.deleteDirectory(rootKeyStorePath)
+        }
     }
 
-    override suspend fun getCoreCrypto(clientId: ClientId?) = withContext(dispatchers.io) {
-        val currentClientId = clientId ?: currentClientIdProvider().fold({ return@withContext Either.Left(it) }, { it })
+    @Suppress("TooGenericExceptionCaught")
+    override suspend fun getCoreCrypto(clientId: ClientId?) = coreCryptoCentralMutex.withLock {
+        withContext(dispatchers.io) {
+            val currentClientId = clientId ?: currentClientIdProvider().fold({ return@withContext Either.Left(it) }, { it })
 
-        val location = "$rootKeyStorePath/${currentClientId.value}".also {
-            // TODO: migrate to okio solution once assert refactor is merged
-            FileUtil.mkDirs(it)
-        }
-        val passphrase = SecurityHelperImpl(passphraseStorage).mlsDBSecret(userId).value
-        return@withContext coreCryptoCentral?.let {
-            Either.Right(it)
-        } ?: run {
-            val cc = coreCryptoCentral(
-                rootDir = "$location/$KEYSTORE_NAME",
-                databaseKey = passphrase
-            )
-            coreCryptoCentral = cc
-            Either.Right(cc)
+            val location = "$rootKeyStorePath/${currentClientId.value}".also {
+                // TODO: migrate to okio solution once assert refactor is merged
+                FileUtil.mkDirs(it)
+            }
+            val passphrase = SecurityHelperImpl(passphraseStorage).mlsDBSecret(userId).value
+            return@withContext coreCryptoCentral?.let {
+                Either.Right(it)
+            } ?: run {
+                val cc = try {
+                    coreCryptoCentral(
+                        rootDir = "$location/$KEYSTORE_NAME",
+                        databaseKey = passphrase
+                    )
+                } catch (e: Exception) {
+                    kaliumLogger.e("Error creating CoreCryptoCentral", e)
+                    kaliumLogger.e("Error creating CoreCryptoCentral stacktrace: ${e.stackTraceToString()}")
```

Comment:
suggestion: let's use `KaliumLogger.logStructuredJson` so we catch it in datadog